edalize/flows/f4pga.py

```diff
@@ -0,0 +1,220 @@
+
+    """Currently supports Xilinx 7-series chips"""
+
+    # F4PGA defined inputs to the underlying tools
+    FLOW_DEFINED_TOOL_OPTIONS = {}
+
+    # Inputs to the F4PGA flow
+    FLOW_OPTIONS = {
+        # Required (used by Yosys)
+        "device": {"type": "str", "desc": "Example: 'artix7'"},
+        # Required (used by Yosys)
+        "part": {"type": "str", "desc": "Example: 'xc7a35tcpg236-1'"},
+        # Required (used by VPR)
+        "chip": {"type": "str", "desc": "Example: 'xc7a50t_test'"},
+        # Optional, defaults to xilinx
+        "arch": {
+            "type": "str",
+            "desc": "Targeting architecture for Yosys. Currently supported is 'xilinx' and that is the default if none specified.",
+        },
+        # Optional, defaults to VPR
+        "pnr_tool": {
+            "type": "str",
+            "desc": "Place and route tool. Valid options are 'vpr'/'vtr' and 'nextpnr'. Defaults to VPR.",
+        },
+    }
+
+    # Standard F4PGA command line arguments to VPR
+    VPR_OPTIONS = [
+        "--max_router_iterations",
+        "500",
+        "--routing_failure_predictor",
+        "off",
+        "--router_high_fanout_threshold",
+        "-1",
+        "--constant_net_method",
+        "route",
+        "--route_chan_width",
+        "500",
+        "--router_heap",
+        "bucket",
+        "--clock_modeling",
+        "route",
+        "--place_delta_delay_matrix_calculation_method",
+        "dijkstra",
+        "--place_delay_model",
+        "delta",
+        "--router_lookahead",
+        "extended_map",
+        "--check_route",
+        "quick",
+        "--strict_checks",
+        "off",
+        "--allow_dangling_combinational_nodes",
+        "on",
+        "--disable_errors",
+        "check_unbuffered_edges:check_route",
+        "--congested_routing_iteration_threshold",
+        "0.8",
+        "--incremental_reroute_delay_ripup",
+        "off",
+        "--base_cost_type",
+        "delay_normalized_length_bounded",
+        "--bb_factor",
+        "10",
+        "--acc_fac",
+        "0.7",
+        "--astar_fac",
+        "1.8",
+        "--initial_pres_fac",
+        "2.828",
+        "--pres_fac_mult",
+        "1.2",
+        "--check_rr_graph",
+        "off",
+        "--suppress_warnings",
+        "sum_pin_class:check_unbuffered_edges:load_rr_indexed_data_T_values:check_rr_node:trans_per_R:check_route:set_rr_graph_tool_comment:calculate_average_switch",
+    ]
+
+    # Creates the flow tree with Yosys and VPR or NextPNR nodes
+    def configure_flow(self, flow_options):
+
+        # Set target
+        # toplevel = self.edam["toplevel"]
+        self.name = self.edam["name"]
+        self.commands.set_default_target(f"{self.name}.bit")
+
+        # Set up nodes
+        synth_tool = "yosys"
+        pnr_tool = "vpr"
+        if "pnr" in flow_options and flow_options.get("pnr_tool") in ["nextpnr"]:
```

Comment:
Is this a typo? Looks like it first checks for `pnr` and then `pnr_tool`. Given my comment above to just change to `pnr` I suggest simplifying it to  `pnr_tool = self.flow_options.get("pnr", "vpr")` to fall back to vpr if no tool is specified, and then

    if not pnr_tool in ["vpr", "nextpnr"]:
        raise(RuntimeError(f"Unsupported P&R tool: '{pnr_tool}'")

I realize now that we should also support setting pnr to `none` if we just want to do synthesis, but we can add that later

Reply:
Should be fixed now in #[48d4d21](https://github.com/olofk/edalize/pull/347/commits/48d4d2182391051a50e31a1fb6933a35211f7334)